### PR TITLE
Propagate context from caller to appropriate places in the code

### PR DESCRIPTION
**Pull request by JoshVanL**
As part of a [wider initiative](https://github.com/dapr/dapr/issues/2478) for dapr to gracefully shutdown (as well as callers having better control of client connections etc), this PR adds contexts to appropriate interface functions so a caller context can be propagated.

Where I could, I have updated packages that used background contexts to instead use the context from the caller. Sometimes this meant using a `closeCh` strategy to catch callers calling `Close()`.

The PR is split into 2 commits and I recommend reviewing them separately. The first changes the actual program code, the second changes unit tests.

I think there is more that can be done in contrib to improve responding to close contexts from the caller, but this is a start.

/cc @artursouza

### CODE REVIEW

**Comment by berndverst** [comment]
> Thanks for this PR. We'll keep it for 1.11 as we're too close to the 1.10 code freeze to make these large changes to interfaces :)

Agreed with Alessandro. As a maintainer this is too last minute given our code freeze effectively end of day, it would also require matching changes in Dapr/Dapr and additional testing. This is too much to coordinate at this time so it will get into 1.11 first thing.

**Comment by JoshVanL** [comment]
Sounds good to me, and also what I was expecting :+1: 

I have a couple inbound dapr PRs based off of this.

**Comment by berndverst** [comment]
The 1.10 release branch has been cut. So this can now be reviewed and eventually merged.

File: pubsub/rabbitmq/rabbitmq.go [hunk 1]
```diff
@@ -563,17 +561,19 @@ func (r *rabbitMQ) reset() (err error) {
 }
 
 func (r *rabbitMQ) isStopped() bool {
-	return r.ctx.Err() != nil
+	select {
+	case <-r.closeCh:
```
Comment (ItalyPaleAle) [inline pubsub/rabbitmq/rabbitmq.go:565]: If the channel is just used for this, consider using an atomic.Bool to use a semaphore. Much simpler than relying on properties like "reading from a closed channel always returns a nil value immediately" :)

File: pubsub/redis/redis.go [hunk 2]
```diff
@@ -399,8 +393,7 @@ func (r *redisStreams) removeMessagesThatNoLongerExistFromPending(ctx context.Co
 
 		// Ack the message to remove it from the pending list.
 		if errors.Is(err, r.client.GetNilValueError()) {
-			// Use the background context in case subscriptionCtx is already closed
```
Comment (ItalyPaleAle) [inline pubsub/redis/redis.go:402]: Please revert this change per the reason explained in the comment :)

File: pubsub/redis/redis.go [hunk 3]
```diff
@@ -252,8 +247,7 @@ func (r *redisStreams) processMessage(msg redisMessageWrapper) error {
 		return err
 	}
 
-	// Use the background context in case subscriptionCtx is already closed
```
Comment (ItalyPaleAle) [inline pubsub/redis/redis.go:255]: Please revert this change per the reason explained in the comment :)

File: state/store.go [hunk 4]
```diff
@@ -23,18 +23,18 @@ import (
 // Store is an interface to perform operations on store.
 type Store interface {
 	BulkStore
-	Init(metadata Metadata) error
-	Features() []Feature
+	Init(ctx context.Context, metadata Metadata) error
+	Features(ctx context.Context) []Feature
```
Comment (ItalyPaleAle) [inline state/store.go:27]: I don't think the Features method should have a context, because it should return a slice synchronously

File: state/postgresql/postgresdbaccess.go [hunk 5]
```diff
@@ -66,7 +66,7 @@ func newPostgresDBAccess(logger logger.Logger) *PostgresDBAccess {
 }
 
 // Init sets up Postgres connection and ensures that the state table exists.
-func (p *PostgresDBAccess) Init(meta state.Metadata) error {
+func (p *PostgresDBAccess) Init(ctx context.Context, meta state.Metadata) error {
 	p.logger.Debug("Initializing Postgres state store")
 
 	p.ctx, p.cancel = context.WithCancel(context.Background())
```
Comment (ItalyPaleAle) [inline state/postgresql/postgresdbaccess.go:72]: Should `ctx` be used here?

File: state/oracledatabase/dbaccess.go [hunk 6]
```diff
@@ -21,7 +21,7 @@ import (
 
 // dbAccess is a private interface which enables unit testing of Oracle Database.
 type dbAccess interface {
-	Init(metadata state.Metadata) error
+	Init(ctx context.Context, metadata state.Metadata) error
 	Ping() error
```
Comment (ItalyPaleAle) [inline state/oracledatabase/dbaccess.go:25]: Ping should probably have a context

File: state/oracledatabase/oracledatabase.go [hunk 7]
```diff
@@ -48,16 +48,16 @@ func newOracleDatabaseStateStore(logger logger.Logger, dba dbAccess) *OracleData
 }
 
 // Init initializes the SQL server state store.
-func (o *OracleDatabase) Init(metadata state.Metadata) error {
-	return o.dbaccess.Init(metadata)
+func (o *OracleDatabase) Init(ctx context.Context, metadata state.Metadata) error {
+	return o.dbaccess.Init(ctx, metadata)
 }
 
 func (o *OracleDatabase) Ping() error {
```
Comment (ItalyPaleAle) [inline state/oracledatabase/oracledatabase.go:55]: Make sure you update all `Ping() error` since you updated the health.Pinger interface

File: state/azure/cosmosdb/cosmosdb.go [hunk 8]
```diff
@@ -191,14 +191,14 @@ func (c *StateStore) Init(meta state.Metadata) error {
 	c.metadata = m
 	c.contentType = m.ContentType
 
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
```
Comment (ItalyPaleAle) [inline state/azure/cosmosdb/cosmosdb.go:194]: Careful not to override `ctx`. Consider naming this something different

File: state/cockroachdb/cockroachdb.go [hunk 9]
```diff
@@ -47,12 +47,12 @@ func internalNew(logger logger.Logger, dba dbAccess) *CockroachDB {
...
-	return c.dbaccess.Init(metadata)
+func (c *CockroachDB) Init(ctx context.Context, metadata state.Metadata) error {
+	return c.dbaccess.Init(ctx, metadata)
 }
 
 // Features returns the features available in this state store.
-func (c *CockroachDB) Features() []state.Feature {
+func (c *CockroachDB) Features(ctx context.Context) []state.Feature {
 	return c.features
 }
```
Comment (ItalyPaleAle) [inline state/cockroachdb/cockroachdb.go:58]: `Ping() error` method below needs a context

File: state/in-memory/in_memory.go [hunk 10]
```diff
@@ -41,34 +41,29 @@ type inMemStateStoreItem struct {
...
-	items map[string]*inMemStateStoreItem
-	lock  *sync.RWMutex
-	log   logger.Logger
-
-	ctx    context.Context
-	cancel context.CancelFunc
+	items   map[string]*inMemStateStoreItem
+	lock    *sync.RWMutex
+	log     logger.Logger
+	closeCh chan struct{}
```
Comment (ItalyPaleAle) [inline state/in-memory/in_memory.go:47]: Here too, a context is safer than a channel

File: state/mysql/mysql.go [hunk 11]
```diff
@@ -297,10 +297,10 @@ func (m *MySQL) ensureStateTable(stateTableName string) error {
 			eTag VARCHAR(36) NOT NULL
 			);`, stateTableName)
 
-		ctx, cancel := context.WithTimeout(context.Background(), m.timeout)
+		//nolint:govet
```
Comment (ItalyPaleAle) [inline state/mysql/mysql.go:300]: Please use a different name for the ctx below rather than a `nolint` directive

File: state/mysql/mysql.go [hunk 12]
```diff
@@ -322,8 +322,8 @@ func schemaExists(db *sql.DB, schemaName string, timeout time.Duration) (bool, e
 	return exists == 1, err
 }
 
-func tableExists(db *sql.DB, tableName string, timeout time.Duration) (bool, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func tableExists(ctx context.Context, db *sql.DB, tableName string, timeout time.Duration) (bool, error) {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
```
Comment (ItalyPaleAle) [inline state/mysql/mysql.go:326]: Don't override the context but use a different name

File: state/mysql/mysql.go [hunk 13]
```diff
@@ -309,8 +309,8 @@ func (m *MySQL) ensureStateTable(stateTableName string) error {
 	return nil
 }
 
-func schemaExists(db *sql.DB, schemaName string, timeout time.Duration) (bool, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func schemaExists(ctx context.Context, db *sql.DB, schemaName string, timeout time.Duration) (bool, error) {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
```
Comment (ItalyPaleAle) [inline state/mysql/mysql.go:313]: Don't override the context but use a different name

File: state/sqlite/sqlite.go [hunk 14]
```diff
@@ -57,7 +57,7 @@ func (s SQLiteStore) GetComponentMetadata() map[string]string {
 }
 
 // Features returns the features available in this state store.
-func (s *SQLiteStore) Features() []state.Feature {
+func (s *SQLiteStore) Features(ctx context.Context) []state.Feature {
 	return []state.Feature{
 		state.FeatureETag,
 		state.FeatureTransactional,
```
Comment (ItalyPaleAle) [inline state/sqlite/sqlite.go:63]: Below, please update `Ping()` toa accept a context too

File: state/cockroachdb/cockroachdb_access.go [hunk 15]
```diff
@@ -81,7 +81,7 @@ func parseMetadata(meta state.Metadata) (*cockroachDBMetadata, error) {
 }
```
Comment (ItalyPaleAle) [inline state/cockroachdb/cockroachdb_access.go:81]: Please update the `Ping()` method below too

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Woah that was a lot of work, so thanks for leading this :)

File: bindings/alicloud/rocketmq/rocketmq.go [hunk 16]
```diff
@@ -35,27 +35,25 @@ type AliCloudRocketMQ struct {
 	settings Settings
 	producer mqw.Producer
 
-	ctx           context.Context
-	cancel        context.CancelFunc
 	backOffConfig retry.Config
+	closeCh       chan struct{}
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/alicloud/rocketmq/rocketmq.go:39]: `Context` really shouldn't be put into structs to be long lived, as per the std lib: "`Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx: `" https://pkg.go.dev/context

I'll add an atomic gate so we don't panic at least.

File: bindings/kafka/kafka.go [hunk 17]
```diff
@@ -29,12 +29,11 @@ const (
...
-	publishTopic    string
-	topics          []string
-	logger          logger.Logger
-	subscribeCtx    context.Context
-	subscribeCancel context.CancelFunc
+	kafka        *kafka.Kafka
+	publishTopic string
+	topics       []string
+	logger       logger.Logger
+	closeCh      chan struct{}
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/kafka/kafka.go:36]: Will do as the above

File: bindings/mqtt3/mqtt.go [hunk 18]
```diff
@@ -40,30 +40,27 @@ type MQTT struct {
 	logger       logger.Logger
 	isSubscribed atomic.Bool
 	readHandler  bindings.Handler
-	ctx          context.Context
-	cancel       context.CancelFunc
 	backOff      backoff.BackOff
+	closeCh      chan struct{}
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:44]: As above

File: bindings/mqtt3/mqtt.go [hunk 19]
```diff
@@ -104,8 +101,8 @@ func (m *MQTT) getProducer() (mqtt.Client, error) {
 	return p, nil
 }
 
-func (m *MQTT) Invoke(parentCtx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
-	producer, err := m.getProducer()
+func (m *MQTT) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	producer, err := m.getProducer(ctx)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:105]: Nvm, they don't have context setup whatsoever yet :sweat_smile:

File: bindings/postgres/postgres.go [hunk 20]
```diff
@@ -59,7 +59,7 @@ func (p *Postgres) Init(metadata bindings.Metadata) error {
 		return errors.Wrap(err, "error opening DB connection")
 	}
 
-	p.db, err = pgxpool.NewWithConfig(context.Background(), poolConfig)
+	p.db, err = pgxpool.NewWithConfig(ctx, poolConfig)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/postgres/postgres.go:62]: This context is scoped to Init, and used by the package to create resources at init time. It doesn't control the lifecycle of the pool. See

https://github.com/jackc/pgx/blob/32c29a6eddb3ba2140bc65c087f71f720160e643/pgxpool/pool.go#L170

https://github.com/jackc/pgx/blob/32c29a6eddb3ba2140bc65c087f71f720160e643/pgxpool/pool.go#L254

https://github.com/jackc/pgx/blob/32c29a6eddb3ba2140bc65c087f71f720160e643/pgxpool/pool.go#L471

I'll add a comment to that affect.

File: middleware/http/bearer/bearer_middleware.go [hunk 21]
```diff
@@ -45,13 +45,13 @@ const (
...
 // GetHandler retruns the HTTP handler provided by the middleware.
-func (m *Middleware) GetHandler(metadata middleware.Metadata) (func(next http.Handler) http.Handler, error) {
+func (m *Middleware) GetHandler(ctx context.Context, metadata middleware.Metadata) (func(next http.Handler) http.Handler, error) {
 	meta, err := m.getNativeMetadata(metadata)
 	if err != nil {
 		return nil, err
 	}
 
-	provider, err := oidc.NewProvider(context.Background(), meta.IssuerURL)
+	provider, err := oidc.NewProvider(ctx, meta.IssuerURL)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply middleware/http/bearer/bearer_middleware.go:54]: This context is used by the OIDC package as the context to fetch the json keys.

https://github.com/coreos/go-oidc/blob/a8ceb9a2043fca2e43518633920db746808b1138/oidc/jwks.go#L227

On the Dapr side, this is called [here](https://github.com/dapr/dapr/blob/24ce7d4ad097a7dc20970d5500dac7441644003a/cmd/daprd/components/middleware_http_bearer.go#L29), which will eventually have a context come from [here](https://github.com/dapr/dapr/blob/3713a058c7c64455ffa8d96d1bc173b7e64f96d2/pkg/components/middleware/http/registry.go#L59), from [here](https://github.com/dapr/dapr/blob/3713a058c7c64455ffa8d96d1bc173b7e64f96d2/pkg/runtime/runtime.go#L694), from [here](https://github.com/dapr/dapr/blob/3713a058c7c64455ffa8d96d1bc173b7e64f96d2/pkg/runtime/runtime.go#L705), [here](https://github.com/dapr/dapr/blob/3713a058c7c64455ffa8d96d1bc173b7e64f96d2/pkg/runtime/runtime.go#L484), which will have the long lived based context that all over dapr runtime will derive from. It should only be cancelled if dapr itself is also shutting down, reloading config etc.

File: middleware/http/wasm/httpwasm.go [hunk 22]
```diff
@@ -164,6 +161,6 @@ func (rh *requestHandler) requestHandler(next http.Handler) http.Handler {
 }
 
 // Close implements io.Closer
-func (rh *requestHandler) Close() error {
+func (rh *requestHandler) Close(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply middleware/http/wasm/httpwasm.go:164]: I see where that is in dapr.

Is this a property we want to maintain? Without a passed context, dapr could hang indefinitely on shutdown/hot reload. Should we add a suitable timeout here instead?

File: pubsub/aws/snssqs/snssqs.go [hunk 23]
```diff
@@ -815,34 +813,31 @@ func (s *snsSqs) Subscribe(subscribeCtx context.Context, req pubsub.SubscribeReq
...
+		<-ctx.Done()
 
 		s.topicsLock.Lock()
 		defer s.topicsLock.Unlock()
 
 		// Remove the handler
 		delete(s.topicHandlers, sanitizedName)
-
-		// If we don't have any topic left, close the poller
-		if len(s.topicHandlers) == 0 {
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pubsub/aws/snssqs/snssqs.go:843]: Thanks, I'll add this logic back in.

I have also added a wait ground to ensure close blocks until all go routines have returned.

File: pubsub/mqtt3/mqtt.go [hunk 24]
```diff
@@ -45,8 +45,7 @@ type mqttPubSub struct {
 	topics          map[string]mqttPubSubSubscription
 	subscribingLock sync.RWMutex
 	reconnectCh     chan struct{}
-	ctx             context.Context
-	cancel          context.CancelFunc
+	closeCh         chan struct{}
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pubsub/mqtt3/mqtt.go:48]: As discussed I like having a channel since we can select over it. Have added an atomic to prevent multiple closes.

Have also added a wait group to block `close()`.

File: state/store.go (hunk 4 above)
Comment (JoshVanL) [reply state/store.go:27]: It needs to accept a context because some of the implementations make network connections.

File: state/azure/cosmosdb/cosmosdb.go (hunk 8 above)
Comment (JoshVanL) [reply state/azure/cosmosdb/cosmosdb.go:194]: Does it matter in this case? We are only using the context on the next Read call.

**Comment by JoshVanL** [comment]
Thank you for the review @ItalyPaleAle! I've addressed or replied to all of your comments.

I'd like to add wait groups to everything so close is graceful, but will do that in a followup PR since this one is already very large.

File: middleware/http/wasm/httpwasm.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply middleware/http/wasm/httpwasm.go:164]: You're welcome to set a timeout but I'd prefer not to alter this. Although it's not specifically part of the interface, it's a standard `io.Closer`...

Maybe we could add a "compile-time assertion" at the top of the files that use this to ensure they comply with io.Closer?

File: state/store.go (hunk 4 above)
Comment (ItalyPaleAle) [reply state/store.go:27]: Which implementations make a network connection in `Features()`? I would say those implementations are buggy and should be changed. Because of how `Features()` is used, IMHO it should be synchronous.

File: state/azure/cosmosdb/cosmosdb.go (hunk 8 above)
Comment (ItalyPaleAle) [reply state/azure/cosmosdb/cosmosdb.go:194]: I would still prefer not to override the variable just in case someone adds more code later and doesn't realize the ctx was overwritten :)

File: state/mysql/mysql.go (hunk 12 above)
Comment (ItalyPaleAle) [reply state/mysql/mysql.go]: Same as above. Just helps preventing psosbile future mistakes :) (I know a person who would probably make those mistakes - me :D )

File: state/store.go (hunk 4 above)
Comment (JoshVanL) [reply state/store.go:27]: You're quite right. I'm not entirely sure where I got that idea from. Think I am going a bit context mad :smile:

File: middleware/http/wasm/httpwasm.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply middleware/http/wasm/httpwasm.go:164]: Fine by me :+1:

[the remark this answers is not in the document]
Comment (JoshVanL) [reply middleware/http/wasm/httpwasm.go]: Added a test to assert `io.Closer`.

File: internal/component/kafka/sasl_oauthbearer.go [hunk 25]
```diff
@@ -107,7 +107,8 @@ func (ts *OAuthTokenSource) Token() (*sarama.AccessToken, error) {
 
 	oidcCfg := ccred.Config{ClientID: ts.ClientID, ClientSecret: ts.ClientSecret, Scopes: ts.Scopes, TokenURL: ts.TokenEndpoint.TokenURL, AuthStyle: ts.TokenEndpoint.AuthStyle}
 
-	timeoutCtx, _ := ctx.WithTimeout(ctx.TODO(), tokenRequestTimeout) //nolint:govet
+	timeoutCtx, cancel := ctx.WithTimeout(ctx.TODO(), tokenRequestTimeout) //nolint:govet
```
Comment (ItalyPaleAle) [inline internal/component/kafka/sasl_oauthbearer.go:110]: Remove the `nolint` directive now

**Comment by berndverst** [comment]
Will the certification tests actually pass? Usually for these kinds of interface changes we need to create a fork of dapr/dapr which brings in this interface change you want to make, make the necessary changes there -- then pin your fork in the various certification tests -- there is a command for this `DAPR_PACKAGE=github.com/myuser/components-contrib@somecommit make replaceruntime-all`

**Comment by berndverst** [comment]
> Will the certification tests actually pass? Usually for these kinds of interface changes we need to create a fork of dapr/dapr which brings in this interface change you want to make, make the necessary changes there -- then pin your fork in the various certification tests -- there is a command for this `DAPR_PACKAGE=github.com/myuser/components-contrib@somecommit make replaceruntime-all`

Aha. As expected your certification tests fail @JoshVanL 

You must now fork dapr/dapr, then import your components contrib version from this very PR here. Then make the necessary changes to get dapr/dapr to build in your fork. Finally run:
`DAPR_PACKAGE=github.com/JoshVanL/dapr@contextfork make replaceruntime-all` and commit all those many `go.mod` and `go.sum` files which changed.

**Comment by berndverst** [comment]
And PLEASE include the interface changes from your other PR (https://github.com/dapr/components-contrib/pull/2499) in that custom dapr/dapr fork (see above). Otherwise we have to do this entire dance multiple times and it slows us down. It also blocks other contributors from making interface changes.

File: internal/component/kafka/sasl_oauthbearer.go (hunk 25 above)
Comment (berndverst) [reply internal/component/kafka/sasl_oauthbearer.go:110]: To be clear, our linter fails because this directive is here

> Error: directive `//nolint:govet` is unused for linter "govet" (nolintlint)

**Comment by JoshVanL** [comment]
> Will the certification tests actually pass? Usually for these kinds of interface changes we need to create a fork of dapr/dapr which brings in this interface change you want to make, make the necessary changes there -- then pin your fork in the various certification tests -- there is a command for this DAPR_PACKAGE=github.com/myuser/dapr@somecommit make replaceruntime-all

Thanks, @ItalyPaleAle are these the steps you were talking about to ping you for?

> You must now fork dapr/dapr, then import your components contrib version from this very PR here. Then make the necessary changes to get dapr/dapr to build in your fork. Finally run:

> And PLEASE include the interface changes from your other PR (https://github.com/dapr/components-contrib/pull/2499) in that custom dapr/dapr fork (see above). Otherwise we have to do this entire dance multiple times and it slows us down. It also blocks other contributors from making interface changes.

This PR is [here](https://github.com/dapr/dapr/pull/5831). I have updated the `go.mod` files to point to the HEAD of that PR. I have also merged in the #2499 and #2498 into this one.

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: bindings/input_binding.go [hunk 26]
```diff
@@ -16,25 +16,29 @@ package bindings
...
 // InputBinding is the interface to define a binding that triggers on incoming events.
 type InputBinding interface {
 	// Init passes connection and properties metadata to the binding implementation.
-	Init(metadata Metadata) error
+	Init(ctx context.Context, metadata Metadata) error
 	// Read is a method that runs in background and triggers the callback function whenever an event arrives.
 	Read(ctx context.Context, handler Handler) error
+	// Close is a method that closes the connection to the binding. Must be
+	// called when the binding is no longer needed to free up resources.
+	io.Closer
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/input_binding.go:32]: Actually because looks like this was not added to other components, perhaps worth not adding it officially here

File: bindings/kafka/kafka.go [hunk 27]
```diff
@@ -110,17 +121,19 @@ func (b *Binding) Read(ctx context.Context, handler bindings.Handler) error {
...
 		// If the component's context has been canceled, do not re-subscribe
-		if b.subscribeCtx.Err() != nil {
+		select {
+		case <-ctx.Done():
 			return
+		default:
 		}
 
-		err := b.kafka.Subscribe(b.subscribeCtx)
+		err := b.kafka.Subscribe(ctx)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/kafka/kafka.go:130]: I see. What is the point of re-subscribing after the read context has been cancelled?

File: bindings/mqtt3/mqtt.go [hunk 28]
```diff
@@ -290,63 +295,66 @@ func (m *MQTT) createClientOptions(uri *url.URL, clientID string) *mqtt.ClientOp
...
-		defer subscribeCancel()
 		select {
 		case <-token.Done():
 			// Subscription went through (sucecessfully or not)
 			err = token.Error()
-		case <-subscribeCtx.Done():
-			err = fmt.Errorf("error while waiting for subscription token: %w", subscribeCtx.Err())
+		case <-time.After(defaultWait):
+			err = errors.New("timed out waiting for subscription to complete")
+		case <-ctx.Done():
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:356]: `connect` is called from `Invoke` and `Read`. Don't we want to return when those contexts are cancelled?

File: bindings/mqtt3/mqtt.go [hunk 29]
```diff
@@ -195,20 +200,20 @@ func (m *MQTT) Read(ctx context.Context, handler bindings.Handler) error {
...
 	if isSubscriber {
-		opts = m.createSubscriberClientOptions(uri, clientID)
+		opts = m.createSubscriberClientOptions(ctx, uri, clientID)
 	} else {
 		opts = m.createClientOptions(uri, clientID)
 	}
 	client := mqtt.NewClient(opts)
 
-	ctx, cancel := context.WithTimeout(m.ctx, defaultWait)
+	ctx, cancel := context.WithTimeout(ctx, defaultWait)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:216]: See above

File: bindings/input_binding.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/input_binding.go:32]: Happy to remove, though I fear that if this is not made part of the interface callers are less likely to call `Close`, or easier to miss because the author didn't attempt to do the type assertion. Ensuring everything is cleaned up and resources released is quite important for things like hot reloading, if we want to do that in future.

File: pubsub/mqtt3/mqtt.go [hunk 30]
```diff
@@ -296,7 +297,7 @@ func (m *mqttPubSub) connect() error {
 }
 
 // Forcefully closes the connection and, after a delay, reconnects
-func (m *mqttPubSub) ResetConection() {
+func (m *mqttPubSub) ResetConection(ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pubsub/mqtt3/mqtt.go:300]: How would the caller cancel this function if it can't provider a context?

**Comment by JoshVanL** [comment]
Thanks for another review @ItalyPaleAle, I have responded to your comments. PTAL

File: pubsub/mqtt3/mqtt.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pubsub/mqtt3/mqtt.go:300]: The connection is in background and should be stopped when the component is stopped

File: bindings/input_binding.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/input_binding.go:32]: Ok, let's leave it.

PS:

> I  fear that if this is not made part of the interface callers are less likely to call Close

The only caller right now is dapr/dapr and in just one spot :) Although we are aware that there's _at least one_ person out there using dapr/components-contrib as a library, it's not officially supported or sanctioned.

File: bindings/kafka/kafka.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/kafka/kafka.go:130]: I believe (please correct me if I'm wrong) that Kafka maintains a single, persistent connection that's started in Init and is used to both publish and subscribe (to all topics). If you tie the lifecycle of that connection to a subscription, cancling the subscription means the connection would be closed and so you'd not be able to publish anymore (and other subscriptions added to the same connection would be dropped)

File: bindings/mqtt3/mqtt.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/mqtt3/mqtt.go:356]: No, we still need the connection to be completed and stored in the object as the persistent connection. Even if an Invoke request is canceled, a connection that's being established should be completed anyways, so the next call to Invoke can use the persistent connection. See how `getProducer` works

File: bindings/kafka/kafka.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/kafka/kafka.go:130]: Thanks @ItalyPaleAle, I guess I didn't totally appreciate the fact that Subscribe restarts the whole kafka client with the new topic. Even so, does it matter? As I understand, calling `Read` multiple times with different handlers isn't supported anyway since we are overriding the existing handlers for the defined topics:

```go
	for _, t := range b.topics {
		b.kafka.AddTopicHandler(t, handlerConfig)
	}
```

My feeling is  we should remove this second `Subscribe` completely. 

The pubsub will be using a different client so we don't need to worry about that one.


I'll switch the context to a `Background`.

File: bindings/mqtt3/mqtt.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:216]: We do because if the parent context is cancelled we want to return straight away without trying to complete the connection. In a shutdown scenario, we want to return straight away, rather than waiting for the connection (indefinitely) or `defaultWait`.

File: pubsub/mqtt3/mqtt.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pubsub/mqtt3/mqtt.go:300]: I think a context is quite important here. If we cannot provide one we could hang indefinitely, even if the very parent context is cancelled of the program.The dapr instance could never stop after a interrupt signal.

File: bindings/mqtt3/mqtt.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/mqtt3/mqtt.go:216]: It does, however... Because if the context times out, `err = ctx.Err()` is non-nil, so we return an error here: `return nil, fmt.Errorf("failed to connect: %w", err)`. This causes `getProducer` not to store the created client:

```go
    p, err := m.connect(ctx, producerClientID, false)
	if err != nil {
		return nil, err
	}
	m.producer = p
```

File: pubsub/mqtt3/mqtt.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pubsub/mqtt3/mqtt.go:300]: TBH we can just delete (or comment-out) this method, which is even unused...

File: bindings/mqtt3/mqtt.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:216]: Is that not the exact behavior we want? The caller should expect that no client was initialized and authed if the given conext was cancelled, and the error tells them that this is the case.

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/mqtt3/mqtt.go:216]: The original intent of the code was to have the connection lazily created. IMHO, a canceled message should not prevent the connection from being created, since it would unnecessarily cause the connection (which could have been established otherwise) to be attempted to be created again on the next invocation. IMHO that seems wasteful.

File: bindings/kafka/kafka.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply bindings/kafka/kafka.go:130]: I did not write that code so I am not sure, but I am happy to accept your proposed change if you can please validate it (essentially try to add and remove topics dynamically and see if that causes any issues)... I think we have a conformance test that does that, but may need some extra work on top.

File: bindings/mqtt3/mqtt.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply bindings/mqtt3/mqtt.go:216]: I disagree that it is wasteful. I argue that is more wasteful to be creating connections, which are an expensive operation, that end up not being used. This is also going to significantly slow down program termination (10 seconds) when the root context is closed before a producer has been created.

In the interest of getting the PR merged I'll remove the context from controlling establishing connections.

**Review by ItalyPaleAle — APPROVED** [review]
LGTM, thanks :)

**Comment by artursouza** [comment]
@JoshVanL Please, resolve conflicts. Thanks.

**Comment by JoshVanL** [comment]
Thanks @artursouza, done

**Comment by JoshVanL** [comment]
@berndverst @ItalyPaleAle Tests look to be passing on the rebase, should be good for another pass :slightly_smiling_face:

**Review by ItalyPaleAle — APPROVED** [review]
